iModelCore/ecobjects/src/ECSchema.cpp

```diff
@@ -3557,9 +3557,22 @@ SchemaWriteStatus ECSchema::WriteToXmlString(WStringR ecSchemaXml, ECVersion ecX
 
     BeXmlWriterPtr xmlWriter = BeXmlWriter::Create();
 
-    SchemaWriteStatus status;
     SchemaXmlWriter schemaWriter(*xmlWriter.get(), *this, ecXmlVersion);
-    if (SchemaWriteStatus::Success != (status = schemaWriter.Serialize()))
+

```

Comment:
It doesn't seem ideal to have this code three times with only minor differences. SchemaWriter::Serialize() may be a better place to deal with this. Also, at this spot here we can only assume the failure was due to the KoQ thing, we do not really know. It may be a different future error which we don't know yet, but still we'd be logging a KoQ problem.

I haven't checked if there is a way to reset BeXmlWriter, but if it's possible, we could handle this inside SchemaWriter::Serialize() at WriteSchemaChild<KindOfQuantity>() so if we get back an error code we can be absolutely sure it is the KoQ problem and nothing else.

Reply:
I had checked this and currently BeXmlWriter doesn't have a way to rest the writer.
Also SchemaWriter has no way to reset the EcXml Version of an object with a setter.
So, I decided to put the fix it at the last common point.
I can maybe add a reset function to BeXmlWriter and a setter to SchemaWriter.

I see your point about future non-koq errors returning FailedToSaveXml and confusing the logs.
I'll try moving the code in Serialise() to resolve this.

Reply:
Had a word with Colin, he prefers to keep the check here itself.
Since the error message already contains the reason for the failure, I've made the new log message more generic by just mentioning that the schema will be re-serialized as ECXml 3.2.